File: 2da Forma/global_response.rb
# Framework: Ruby on Rails
# Archivo: global_response.rb

module Responses
  class GlobalResponse
    attr_reader :success, :message, :data, :errors, :pagination

    def initialize(success:, message:, data: nil, errors: nil, pagination: nil)
      @success = success
      @message = message
      @data = data
      @errors = errors
      @pagination = pagination
    end

    # Success responses
    def self.ok(message = 'Operación exitosa')
      new(success: true, message: message)
    end

    def self.ok_with_data(data, message = 'Operación exitosa')
      new(success: true, message: message, data: data)
    end

    def self.ok_paginated(data, total_items:, current_page:, page_size:, message: 'Datos obtenidos exitosamente')
      pagination = PaginationInfo.new(
        total_items: total_items,
        current_page: current_page,
        page_size: page_size
      )
      new(success: true, message: message, data: data, pagination: pagination)
    end

    # Error responses
    def self.fail(message)
      new(success: false, message: message)
    end

    def self.fail_with_errors(message, errors)
      new(success: false, message: message, errors: errors)
    end

    def self.validation_error(message, validation_errors)
      new(success: false, message: message, errors: validation_errors)
    end

    def self.not_found(message = 'Recurso no encontrado')
      new(success: false, message: message)
    end

    def self.unauthorized(message = 'No autorizado')
      new(success: false, message: message)
    end

    def self.server_error(message = 'Error interno del servidor')
      new(success: false, message: message)
    end

    def to_json(*_args)
      {
        success: @success,
        message: @message,
        data: @data,
        errors: @errors,
        pagination: @pagination&.to_hash
      }.compact.to_json
    end

    def as_json(options = {})
      {
        success: @success,
        message: @message,
        data: @data,
        errors: @errors,
        pagination: @pagination&.as_json
      }.compact
    end
  end

  class PaginationInfo
    attr_reader :total_items, :current_page, :page_size, :total_pages, :has_previous, :has_next

    def initialize(total_items:, current_page:, page_size:)
      @total_items = total_items
      @current_page = current_page
      @page_size = page_size
      @total_pages = (total_items.to_f / page_size).ceil
      @has_previous = current_page > 1
      @has_next = current_page < @total_pages
    end

    def to_hash
      {
        total_items: @total_items,
        current_page: @current_page,
        page_size: @page_size,
        total_pages: @total_pages,
        has_previous: @has_previous,
        has_next: @has_next
      }
    end

    def as_json(options = {})
      to_hash
    end
  end
end

File: 2da Forma/GlobalResponse.php
<?php
// Framework: Laravel
// Archivo: GlobalResponse.php

namespace App\Http\Responses;

use Illuminate\Contracts\Support\Arrayable;
use Illuminate\Contracts\Support\Jsonable;
use JsonSerializable;

class GlobalResponse implements Arrayable, Jsonable, JsonSerializable
{
    protected bool $success;
    protected string $message;
    protected mixed $data;
    protected mixed $errors;
    protected ?PaginationInfo $pagination;

    private function __construct(
        bool $success,
        string $message,
        mixed $data = null,
        mixed $errors = null,
        ?PaginationInfo $pagination = null
    ) {
        $this->success = $success;
        $this->message = $message;
        $this->data = $data;
        $this->errors = $errors;
        $this->pagination = $pagination;
    }

    // Success responses
    public static function ok(string $message = 'Operación exitosa'): self
    {
        return new self(true, $message);
    }

    public static function okWithData(mixed $data, string $message = 'Operación exitosa'): self
    {
        return new self(true, $message, $data);
    }

    public static function okPaginated(
        mixed $data,
        int $totalItems,
        int $currentPage,
        int $pageSize,
        string $message = 'Datos obtenidos exitosamente'
    ): self {
        $pagination = new PaginationInfo($totalItems, $currentPage, $pageSize);
        return new self(true, $message, $data, null, $pagination);
    }

    // Error responses
    public static function fail(string $message): self
    {
        return new self(false, $message);
    }

    public static function failWithErrors(string $message, mixed $errors): self
    {
        return new self(false, $message, null, $errors);
    }

    public static function validationError(string $message, array $validationErrors): self
    {
        return new self(false, $message, null, $validationErrors);
    }

    public static function notFound(string $message = 'Recurso no encontrado'): self
    {
        return new self(false, $message);
    }

    public static function unauthorized(string $message = 'No autorizado'): self
    {
        return new self(false, $message);
    }

    public static function serverError(string $message = 'Error interno del servidor'): self
    {
        return new self(false, $message);
    }

    public function toArray(): array
    {
        $result = [
            'success' => $this->success,
            'message' => $this->message,
        ];

        if ($this->data !== null) {
            $result['data'] = $this->data;
        }

        if ($this->errors !== null) {
            $result['errors'] = $this->errors;
        }

        if ($this->pagination !== null) {
            $result['pagination'] = $this->pagination->toArray();
        }

        return $result;
    }

    public function toJson($options = 0): string
    {
        return json_encode($this->jsonSerialize(), $options);
    }

    public function jsonSerialize(): array
    {
        return $this->toArray();
    }
}

class PaginationInfo implements Arrayable, JsonSerializable
{
    public int $totalItems;
    public int $currentPage;
    public int $pageSize;
    public int $totalPages;
    public bool $hasPrevious;
    public bool $hasNext;

    public function __construct(int $totalItems, int $currentPage, int $pageSize)
    {
        $this->totalItems = $totalItems;
        $this->currentPage = $currentPage;
        $this->pageSize = $pageSize;
        $this->totalPages = (int) ceil($totalItems / $pageSize);
        $this->hasPrevious = $currentPage > 1;
        $this->hasNext = $currentPage < $this->totalPages;
    }

    public function toArray(): array
    {
        return [
            'totalItems' => $this->totalItems,
            'currentPage' => $this->currentPage,
            'pageSize' => $this->pageSize,
            'totalPages' => $this->totalPages,
            'hasPrevious' => $this->hasPrevious,
            'hasNext' => $this->hasNext,
        ];
    }

    public function jsonSerialize(): array
    {
        return $this->toArray();
    }
}